### ISSUE TITLE
Webpdf export doesn't seem to work in K8s/JupyterHub/JupyterLab setting

### ISSUE THREAD

**Issue by athornton** [jupyter/nbconvert#1434]
I can't find a way to make Export As Webpdf work in my JH/JL/K8s setting.

Specifically: if I haven't installed `nbconvert[webpdf]` I still have the JL menu item, but it tells me to install `nbconvert[webpdf]`.  OK, fair, although I would prefer that I not have a menu item that doesn't work.

So I install it.  Then I get that I have to use `allow-chromium-download`.  Well, that's a problem, since I don't see how to add that flag to whatever the menu item invokes.  But, OK, first I tried installing chromium-headless into my container.  That didn't help but the container is Centos7-based so I'm totally willing to believe I need a newer Chromium.  So, fine, I run pyppeteer-install, and it downloads chromium into my .local.  But *then* when I try "export as webpdf", I get a pause for a while and then a new page that says `nbconvert failed: Browser closed unexpectedly:` .  My suspicion, although I have no evidence for this, is that the difference in the URL to the rendered PDF between what the user's browser sees and what the Jupyter environment sees, is what's breaking this.  Presumably it's been tested to death on local JL installations, and this wouldn't be the first time that I've been bitten by the assumption that the browser and the Jupyter server see the same URL.

I don't actually need this to work; if I can find out where the menu item is being inserted into the export menu, and just *not add it* I'm perfectly happy with my existing Export as PDF functionality.  But it bothers me that I have a menu item that doesn't work, and no way that I can find to not show it to my users.  So...where is that menu item injected into JupyterLab, and how do I make it not do that?

Pick a notebook.  "Export as webpdf".

**6.0.6**

**Comment by ghuname** [comment]
Try to find out where pyppeteer is installed and than edit louncher.py.
In my case louncher.py is located at /opt/conda/lib/python3.8/site-packages/pyppeteer/launcher.py

In louncher.py at the very beginning you will see:

DEFAULT_ARGS = [
    '--disable-background-networking',
    '--disable-background-timer-throttling',
    '--disable-breakpad',
    '--disable-browser-side-navigation',
    '--disable-client-side-phishing-detection',
    '--disable-default-apps',
    '--disable-dev-shm-usage',
    '--disable-extensions',
    '--disable-features=site-per-process',
    '--disable-hang-monitor',
    '--disable-popup-blocking',
    '--disable-prompt-on-repost',
    '--disable-sync',
    '--disable-translate',
    '--metrics-recording-only',
    '--no-first-run',
    '--safebrowsing-disable-auto-update',
    '--enable-automation',
    '--password-store=basic',
    '--use-mock-keychain',
    '--no-sandbox'
]

Original DEFAULT_ARGS is with no last parameter '--no-sandbox'.
Just add that parameter and it will work.

If it doesn't work than do the following:
sudo apt-get update
sudo apt-get install gconf-service libasound2 libatk1.0-0 libatk-bridge2.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libexpat1 libfontconfig1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libnspr4 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 libxtst6 ca-certificates fonts-liberation libappindicator1 libnss3 lsb-release xdg-utils wget libcairo-gobject2 libxinerama1 libgtk2.0-0 libpangoft2-1.0-0 libthai0 libpixman-1-0 libxcb-render0 libharfbuzz0b libdatrie1 libgraphite2-3 libgbm1

Now it must work.

**Comment by yuvipanda** [comment]
`--no-sandbox` does sound a bit scary though - it means clientside JS could possibly have server-side execution. However, if you're actively converting *running* notebooks into PDF on your JupyterHub, it's probably fine since that JS can already run arbitrary code on the server :D Gotta be careful though.

**Comment by yuvipanda** [comment]
@athornton @ghuname https://github.com/jupyter/nbconvert/pull/1516 should help. I am defaulting it to 'off' though, so would need explicit enabling to work.

**Comment by athornton** [comment]
Explicit enabling doesn't seem to work (I'm just now getting back to this, 14 months later, yes).

Here's the edited trace:

```
[athornton@nb-athornton ~]$ jupyter nbconvert ~/play/Visualization.ipynb  --to webpdf --debug[NbConvertApp] Searching ['/home/athornton', '/home/athornton/.jupyter', '/home/athornton/.local/etc/jupyter', '/opt/lsst/software/stack/conda/miniconda3-py38_4.9.2/envs/lsst-scipipe-3.0.0/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files
[NbConvertApp] Looking for jupyter_config in /etc/jupyter
[NbConvertApp] Looking for jupyter_config in /usr/local/etc/jupyter
[NbConvertApp] Looking for jupyter_config in /opt/lsst/software/stack/conda/miniconda3-py38_4.9.2/envs/lsst-scipipe-3.0.0/etc/jupyter
[NbConvertApp] Looking for jupyter_config in /home/athornton/.local/etc/jupyter
[NbConvertApp] Looking for jupyter_config in /home/athornton/.jupyter
[NbConvertApp] Looking for jupyter_config in /home/athornton
[NbConvertApp] Looking for jupyter_nbconvert_config in /etc/jupyter
[NbConvertApp] Looking for jupyter_nbconvert_config in /usr/local/etc/jupyter
[NbConvertApp] Loaded config file: /usr/local/etc/jupyter/jupyter_nbconvert_config.py

...

[NbConvertApp] Building PDF
[INFO] Starting Chromium download.
100%|██████████████████████████████████████████████████████████████████████████████████████| 109M/109M [00:00<00:00, 190Mb/s]
[INFO] Beginning extraction
[INFO] Chromium extracted to: /home/athornton/.local/share/pyppeteer/local-chromium/588429
Traceback (most recent call last):
  File "/opt/lsst/software/stack/conda/miniconda3-py38_4.9.2/envs/lsst-scipipe-3.0.0/bin/jupyter-nbconvert", line 10, in <module>
    sys.exit(main())

...

    browser = await self._check_launch_reqs()(
  File "/opt/lsst/software/stack/conda/miniconda3-py38_4.9.2/envs/lsst-scipipe-3.0.0/lib/python3.8/site-packages/pyppeteer/launcher.py", line 307, in launch
    return await Launcher(options, **kwargs).launch()
  File "/opt/lsst/software/stack/conda/miniconda3-py38_4.9.2/envs/lsst-scipipe-3.0.0/lib/python3.8/site-packages/pyppeteer/launcher.py", line 168, in launch
    self.browserWSEndpoint = get_ws_endpoint(self.url)
  File "/opt/lsst/software/stack/conda/miniconda3-py38_4.9.2/envs/lsst-scipipe-3.0.0/lib/python3.8/site-packages/pyppeteer/launcher.py", line 227, in get_ws_endpoint
    raise BrowserError('Browser closed unexpectedly:\n')
pyppeteer.errors.BrowserError: Browser closed unexpectedly:
```

The contents of `/usr/local/etc/jupyter/jupyter_nbconvert_config.py` are:

```
c.WebPDFExporter.allow_chromium_download = True
c.WebPDFExporter.disable_sandbox = True
```

**Comment by jaguillette** [comment]
I found this discussion in doing my own research to get the webpdf export working in a containerized environment, and I believe I've found a way to consistently have the `--no-sandbox` option used. I used the [notebook config](https://jupyter-notebook.readthedocs.io/en/stable/config.html) and the [nbconvert config](https://nbconvert.readthedocs.io/en/latest/config_options.html) to set `c.WebPDFExporter.disable_sandbox = True` in each.

Basically, I needed a `~/.jupyter/jupyter_nbconvert_config.py` file and a `~/.jupyter/jupyter_notebook_config.py` for the user running juypter (notebook, in my case), both of them containing the line 

```python
c.WebPDFExporter.disable_sandbox = True
```

With the installations you described and those config files in place, the webpdf export works in the Docker container I'm working with. Hopefully that works for you, too.
